lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs

```diff
@@ -139,21 +138,67 @@ withNetworkLayer tr project k = k NetworkLayer
-    parseBlockHeader blockHash =
-        case fromText (BF.unBlockHash blockHash) of
-            Right hash -> pure hash
-            Left tde -> throwError $ InvalidBlockHash blockHash tde
+class FromBlockfrost b w where
+    fromBlockfrost :: forall m. MonadError BlockfrostError m => b -> m w
+
+instance FromBlockfrost BF.Block BlockHeader where
+    fromBlockfrost block@BF.Block{..} = do
+        slotNo <- case _blockSlot of
+            Just s -> pure $ SlotNo $ fromIntegral $ BF.unSlot s
+            Nothing -> throwError $ NoSlotError block
+        blockHeight <- case _blockHeight of
+            Just height -> pure $ Quantity $ fromIntegral height
+            Nothing -> throwError $ NoBlockHeight block
+        headerHash <- parseBlockHeader _blockHash
+        parentHeaderHash <- for _blockPreviousBlock parseBlockHeader
+        pure BlockHeader { slotNo, blockHeight, headerHash, parentHeaderHash }
+      where
+        parseBlockHeader blockHash =
+            case fromText (BF.unBlockHash blockHash) of
+                Right hash -> pure hash
+                Left tde -> throwError $ InvalidBlockHash blockHash tde
+
+instance FromBlockfrost BF.ProtocolParams ProtocolParameters where
+    fromBlockfrost BF.ProtocolParams{..} = do
+        decentralizationLevel <-
+            let percentage = mkPercentage $
+                    toRational _protocolParamsDecentralisationParam
+            in case percentage of
+                Left PercentageOutOfBoundsError ->
+                    throwError $ InvalidDecentralizationLevelPercentage
+                        _protocolParamsDecentralisationParam
+                Right level -> pure $ DecentralizationLevel level
+        let intToQuantity :: (Num q, Integral i) => i -> Quantity s q
+            intToQuantity = Quantity . fromIntegral
+            txParameters = TxParameters
+                { getFeePolicy =  LinearFee
+                    (intToQuantity _protocolParamsMinFeeA)
+                    (intToQuantity _protocolParamsMinFeeB)
+                , getTxMaxSize =
+                    intToQuantity _protocolParamsMaxTxSize
+                , getTokenBundleMaxSize =
+                    TokenBundleMaxSize $ TxSize $ fromIntegral $
+                        BF.unQuantity _protocolParamsMaxValSize
+                , getMaxExecutionUnits =
+                    ExecutionUnits
+                        { executionSteps = fromIntegral $
+                            BF.unQuantity _protocolParamsMaxTxExSteps
+                        , executionMemory = fromIntegral $
+                            BF.unQuantity _protocolParamsMaxTxExMem
+                        }
+                }
+            desiredNumberOfStakePools = fromIntegral _protocolParamsNOpt
+            minimumUTxOvalue = MinimumUTxOValueCostPerWord $ Coin $
+                fromIntegral _protocolParamsCoinsPerUtxoWord
+            stakeKeyDeposit = Coin $ fromIntegral _protocolParamsKeyDeposit
```

Comment:
Haven't looked at what the types are here, but you could consider using [`intCast`](https://hackage.haskell.org/package/int-cast-0.1.0.0/candidate/docs/Data-IntCast.html) instead of `fromIntegral` to make it clear that there can be no over- or underflow. It's not yet used a lot in the wallet, but it is used in a few places, and seems practical.

Reply:
Thanks for the suggestion, I have refactored integral conversions using `intCast` where applicable. Please re-review.